### DISCUSSION TITLE
How to onboard Identity Provider Certificate into Keys

### DISCUSSION THREAD

**Discussion by vishrantgupta** [keycloak/keycloak#27961] [Q&A]
I have an Identity Provider Certificate and I am trying to onboard it into Keys, but I don't see an option to import it unless the Private Key is specified under "Add keystore" -> "RSA". This is useful to validate the access token.

The Identity Provider does not share the Private Key example Auth0.

**Comment by justin-tay** [comment]
Do you mean to validate the signature on the ID Token?

In the [OpenID Connect v1.0 identity providers](https://www.keycloak.org/docs/latest/server_admin/index.html#_identity_broker_oidc) configuration you will need to enable the `Validate Signatures` option. Ideally the Identity Provider will expose a JWKS URL for Keycloak to get the public keys, but otherwise you can configure the `Validating public key`.

![validate](https://github.com/keycloak/keycloak/assets/49700559/31138d54-ac43-4b6a-8b1a-bfefeef31df5)

**Reply by bbartels** [reply]
Once that is enabled, should the key be exposed via: `https://{HELIX_URL}/auth/realms/{REALM}/protocol/openid-connect/certs`?
Because that is not the case for me once i've enabled validate signatures

**Reply by justin-tay** [reply]
Don't understand your question. This is for providing the public key to validate tokens from the identity provider that keycloak is brokering. Why would the key be exposed?

**Reply by bbartels** [reply]
Because the gocloak client library checks the /certs endpoint for a kid matching the one that the jwt was signed by. In order for this check to pass it would have to expose the key of the external Identity Provider at `https://{HELIX_URL}/auth/realms/{REALM}/protocol/openid-connect/certs` https://github.com/Nerzal/gocloak/blob/main/client.go#L485.
Or is this something that isn't supported with keycloak?

I am trying to achieve this scenario: https://github.com/keycloak/keycloak/discussions/37396

**Reply by justin-tay** [reply]
The tokens you will be getting are issued by keycloak and not from the brokered identity provider. Are you expecting the original token? As in internal to external token exchange?

**Reply by bbartels** [reply]
I want to enable a scenario where a token issued by a the external IDP is accepted by the application who has keycloak as an IDP. So when they call the gocloak client library that it successfully validates the externally issued token.
As the gocloak library first does a cert check, this fails, as the token was signed by the external idp.
Is this a scenario keycloak allows for?

**Reply by justin-tay** [reply]
I reread your scenario a few times but I don't understand it. How are you getting the externally issued token? Directly from the non-keycloak authorization server? As in using client credential flow? If you are getting the token from the non-keycloak authorization server your client application should be pointing to the JWKS url of that authorization server to validate the token? Why do you need to pass the externally issued token to your application instead of one issued by keycloak?

**Reply by bbartels** [reply]
Apologies, I was being confusing. Here is the full context:
Primary IDP -> issues access tokens via a custom flow (kubernetes service identity)
Keycloak -> IDP of vendor product we are trying to integrate

The vendor product has a hard dependency on keycloak. We want to leverage our internal custom flow in order to fetch access tokens from our primary IDP to interact with the vendor product. As there is no way to configure the vendor product to just bypass keycloak and use our primary IDP, we are required to use keycloak.

There is two approaches to make this work in a transparent way:

1. Configure keycloak in a way that when an access token issued by the primary idp can be verified by the vendor product as a valid one
    a. Configure keycloak so it publishes the JWKS from our primary IDP at `https://{KEYCLOAK_URL}/auth/realms/{REALM}/protocol/openid-connect/certs` so the vendor product _thinks_ the token was signed by keycloak (even though it wasn't)
    b. Make the vendor product call out to keycloak via introspect and have keycloak validate it.
    c. Exchange externally issued JWT for keycloak internal one (Would cause significant auth friction)
2. Make a change to the vendor product to get rid of keycloak as a hard dependency and allow our primary IDP to be used directly.

This is our scenario, we are required to use tokens from our primary IDP in order to talk to the vendor product. We can't exchange an external token to an internal one because that would add a lot of complexity over how auth works in our company in general.
I couldn't figure out from the documentation whether keycloak supports `1.a`, I thought configuring keycloak to use an external identity provider would result in the external providers JWKS being publish via keycloak. But I may be misunderstanding.

In your opinion, which of the three approaches (1.a, 1.b, 1.c, 2) would be the most sensible?

**Reply by justin-tay** [reply]
The feature described in this thread is totally not relevant to your use case.

If all that is required is that the vendor application needs to validate the signature of the access token, then you should just do 1a. and configure the key via `Realm settings > Keys` and add a `java-keystore` provider with the key.

If the vendor application is actually managing users, clients, realm, authorization etc in keycloak and is expecting certain claims in the access token then I don't really see how this will work. This vendor application doesn't have any fine grain authorization? Just passing signature verification of the JWT access token will do?

**Reply by bbartels** [reply]
It doesn't yet have fine grained access control. The proposed implementation will take the `sub` claim of the supplied JWT and then reach out to keycloak to retrieve information about user controls. Which should work in this scenario.

The java-keystore presumably wouldn't rotate if the remote JWKS rotates, right? So we'd have to do this on a schedule?

**Reply by justin-tay** [reply]
If the JWKS url on the vendor app is customizable you can write an endpoint that combines both JWKS. Alternatively you can write a custom `KeyProvider` for keycloak but i think it might be more of a hassle to maintain.

**Reply by bbartels** [reply]
Okay, perfect, thank you! This has helped a ton :) I'll have a think as to what the best way to proceed on this would be!

**Reply by bbartels** [reply]
Out of curiosity, what is the keycloak recommended way to go about this? Exchange an external for an internal token? Can keycloak introspection delegate to the introspection of an external IDP?

**Reply by justin-tay** [reply]
Your client should be getting the access token from keycloak using client credentials flow to pass to your vendor app.

**Reply by bbartels** [reply]
So there is no "recommended" flow for authing with an external IDP and using those creds with keycloak in some way? As in, what I am trying to do is outside of the parameters keycloak intends to operate in?
The only recommended way of doing things is via keycloak native client credential flow?

**Reply by justin-tay** [reply]
You can ask around but from what I can tell your options are like public client external to internal token exchange and I'm not sure if you can do public client token exchange in keycloak going forward. If you need a confidential client to do the token exchange I think that defeats the purpose. The other possibility is if your external idp would support exchanging the access token for a jwt client assertion that you can then use with keycloak to get the access token.
